frappe/workflow/doctype/workflow_action/workflow_action.py

```diff
@@ -132,20 +138,110 @@ def return_link_expired_page(doc, doc_workflow_state):
+		clear_old_workflow_actions_using_role(doc)
+	else:
+		# backwards compatibility
+		# for workflow actions saved using user
+		clear_old_workflow_actions_using_user(doc, user)
+		update_completed_workflow_actions_using_user(doc, user)
+
+def get_allowed_roles(user, workflow, workflow_state):
 	user = user if user else frappe.session.user
-	frappe.db.delete("Workflow Action", {
-		"reference_doctype": doc.get("doctype"),
-		"reference_name": doc.get("name"),
-		"user": ("!=", user),
-		"status": "Open"
-	})
 
-def update_completed_workflow_actions(doc, user=None):
+	allowed_roles = frappe.get_all('Workflow Transition',
+		fields='allowed',
+		filters=[['parent', '=', workflow],
+		['next_state', '=', workflow_state]],
+		pluck = 'allowed')
+
+	user_roles = set(frappe.get_roles(user))
+	return set(allowed_roles).intersection(user_roles)
+
+def get_workflow_action_by_role(doc, allowed_roles):
+	WorkflowAction = DocType("Workflow Action")
+	return frappe.qb.from_(WorkflowAction).select(WorkflowAction.name).where(
+			WorkflowAction.status == 'Open',
+		).where(
+			WorkflowAction.role.isin(list(allowed_roles))
+		).where(
+			WorkflowAction.reference_doctype == doc.get('doctype')
+		).where(
+			WorkflowAction.reference_name == doc.get('name')
+		).where(
+			WorkflowAction.is_deleted == 0,
+		).orderby('role').limit(1).run()
+
+def update_completed_workflow_actions_using_role(doc, user=None, allowed_roles = set(), workflow_action_name=None):
 	user = user if user else frappe.session.user
-	frappe.db.sql("""UPDATE `tabWorkflow Action` SET `status`='Completed', `completed_by`=%s
-		WHERE `reference_doctype`=%s AND `reference_name`=%s AND `user`=%s AND `status`='Open'""",
-		(user, doc.get('doctype'), doc.get('name'), user))
+	WorkflowAction = DocType("Workflow Action")
+
+	if not workflow_action_name or not workflow_action_name[0]:
+		return
+
+	frappe.qb.update(WorkflowAction).set(
+		WorkflowAction.status, 'Completed',
+	).set(
+		WorkflowAction.completed_by, user
+	).where(
+		WorkflowAction.name == workflow_action_name[0][0],
+	).run()
+
+def clear_old_workflow_actions_using_role(doc):
+	WorkflowAction = DocType("Workflow Action")
+	frappe.qb.update(WorkflowAction).set(WorkflowAction.is_deleted, 1).where(
+		WorkflowAction.reference_doctype == doc.get("doctype")
+	).where(
+		WorkflowAction.reference_name == doc.get("name")
+	).where(
+		WorkflowAction.status == "Open"
+	).run()
+
+def clear_old_workflow_actions_using_user(doc, user=None):
+	user = user if user else frappe.session.user
+	if not frappe.db.has_column('Workflow Action', 'user'):
+		return False
+	WorkflowAction = DocType("Workflow Action")
+	frappe.qb.update(WorkflowAction).set(
+		WorkflowAction.is_deleted, 1
+	).where(
+		WorkflowAction.reference_doctype == doc.get("doctype")
+	).where(
+		WorkflowAction.reference_name == doc.get("name")
+	).where(
+		WorkflowAction.user != user
+	).where(
+		WorkflowAction.status == "Open"
+	).run()
```

Comment:
1. This and the following function seem to Falsy values if the user column exists or not. 
2. Should the user column be used at all? If the schema files don't define the Workflow Action DocType with a user table, why maintain the column at all? This column is one `bench trim-tables` away from being deleted. If it has information we need, transform and move it into the role column

Reply:
> 1. This and the following function seem to Falsy values if the user column exists or not.
This function would never be called if the user column does not exist. The if condition at the top is probably needed. I added it to make sure the function checks if it should run or not by itself. I think the if condition at the top should be removed.

> 2. Should the user column be used at all? If the schema files don't define the Workflow Action DocType with a user table, why maintain the column at all? This column is one `bench trim-tables` away from being deleted. If it has information we need, transform and move it into the role column
It is not possible to figure out which role gave the user the permission to perform an action just by looking at the user. We do not even store the previous state so no way to know how the document came to the state it is right now.